easybuild/easyblocks/generic/gopackage.py

```diff
@@ -0,0 +1,169 @@
+
+    def build_step(self):
+        """If Go package is not native go module, lets try to make the module."""
+
+        go_mod_file = os.path.join('./', 'go.mod')
+        go_sum_file = os.path.join('./', 'go.sum')
+
+        if not os.path.exists(go_mod_file) or not os.path.isfile(go_mod_file):
+            self.log.warn("go.mod not found! This is not natively supported go module. Trying to init module.")
+
+            if self.cfg['modulename'] is None:
+                raise EasyBuildError("Installing non-native go module. You need to specify 'modulename' in easyconfig")
+
+            # for more information about migrating to go modules
+            # see: https://blog.golang.org/migrating-to-go-modules
+
+            # go mod init
+            cmd = ' '.join([self.go_cmd, 'mod init', self.cfg['modulename']])
+            run_cmd(cmd, log_all=True, simple=True)
+
+            if self.cfg['forced_deps']:
+                for dep in self.cfg['forced_deps']:
+                    # go get specific dependencies which locks them in go.mod
+                    cmd = ' '.join([self.go_cmd, 'get %s@%s' % dep])
+                    run_cmd(cmd, log_all=True, simple=True)
+
+            # go build ./...
+            cmd = ' '.join([self.go_cmd, 'build ./...'])
+            run_cmd(cmd, log_all=True, simple=True)
+
+            # go test ./...
+            cmd = ' '.join([self.go_cmd, 'test ./...'])
+            run_cmd(cmd, log_all=True, simple=True)
+
+            # go mod tidy
+            cmd = ' '.join([self.go_cmd, 'mod tidy'])
+            run_cmd(cmd, log_all=True, simple=True)
+
+            # go build ./... again
+            cmd = ' '.join([self.go_cmd, 'build ./...'])
+            run_cmd(cmd, log_all=True, simple=True)
+
+            # go test ./... again
+            cmd = ' '.join([self.go_cmd, 'test ./...'])
+            run_cmd(cmd, log_all=True, simple=True)
+
+            self.log.warn('Include generated go.mod and go.sum via patch to ensure locked dependencies '
+                          'and run this easyconfig again.')
+            cmd = 'cat go.mod'
+            run_cmd(cmd, log_all=True, simple=True)
+
+            cmd = 'cat go.sum'
+            run_cmd(cmd, log_all=True, simple=True)
+
+        if not os.path.exists(go_sum_file) or not os.path.isfile(go_sum_file):
+            raise EasyBuildError("go.sum not found! This module has no locked dependency versions.")
+
+    def install_step(self):
+        """Install Go package to a custom path"""
+
+        # actually install Go package
+        cmd = ' '.join(
+            [self.cfg['preinstallopts'], self.go_cmd, 'install', self.cfg['installopts']])
+        run_cmd(cmd, log_all=True, log_ok=True, simple=True)
+
+    def sanity_check_step(self):
+        """Custom sanity check for Go package."""
+
+        # Go Package should produce something into bin directory
+        custom_paths = {
+            'files': [],
+            'dirs': ['bin'],
```

Comment:
This is very generic... We can't derive anything from software name?

Reply:
I'm not very confident in naming in go. But it should match last part of Go module name, considering that main.go is in root. Otherwise it could be any name from any sub-folder producing binary.

The main reason for this "basic" check was that for example compiling gonum doesn't produce any binary, since it is intended to be used as dependency in go source. So you are able to build and install this package, but no binary will be dropped in bin and nothing you can use on easybuild will be installed.

Should we check the name (namelower)?
Consider this: https://github.com/evolbioinfo/gotree
They have 3 different names in single paragraph: 
`Gotree` in heading
`GoTree` just below in description
`gotree` as cli command

Reply:
I think assuming that there will be a `bin/%(namelower)s` makes sense, since this is only a default (it's easy to override in the easyconfig file).

This would work for both `goalign` and `gotree`.

Maybe also add `%(namelower)s --help` as default sanity check commands (via `custom_commands` here)?

Also easy to override in easyconfig file via `sanity_check_commands`.